Blog: Apache DataFusion is now the fastest single node engine for querying Apache Parquet files (#33)
apache/datafusion-site: _posts/2024-11-18-datafusion-fastest-single-node-parquet-clickbench.md

Best in class performance on Parquet is now available to anyone. DataFusion’s
open design lets you start quickly with a full featured Query Engine, including
SQL, data formats, catalogs, and more, and then customize any behavior you need.
I predict the continued emergence of new classes of data systems now that
creators can focus the bulk of their innovation on areas such as query
languages, system integrations, and data formats rather than trying to play
catchup with core engine performance.

ClickBench also includes results for proprietary storage formats, which require
costly load / export steps, making them useful in fewer use cases and thus much
less important than open formats (though the idea of use case specific formats
is interesting[^2]).

This blog post highlights some of the techniques we used to achieve this
performance, and celebrates the teamwork involved.

# Rallying The Community around Performance

> In July, 2024 [Mehmet Ozan Kabak], CEO of [Synnada], [called on the community to

Author: I certainly was inspired by the comment to help me focus where I spent my time reviewing PRs and helping push them through, though it is a good point that this may imply it motivated others as well, when I don't really know what did. 

Perhaps we could rephrase the motivation with something like this?

"Performance has long been a focus for DataFusion: one of the core benefits of DataFusion is its core performance, which  both excites contributors and attracts users. There seems to have been a renewed focus on performance recently, including a call in July 2024 from Mehmet ....?

> In July, 2024 [Mehmet Ozan Kabak], CEO of [Synnada], [called on the community to

Reviewer: Sounds good to me

> In July, 2024 [Mehmet Ozan Kabak], CEO of [Synnada], [called on the community to

Author: I rephrased in 2507e67

improve DataFusion’s performance]. This got many of us excited (who doesn’t love
a challenge!), and we have subsequently rallied to steadily improve the
performance release on release as shown in Figure 2.

> [Mehmet Ozan Kabak]: https://www.linkedin.com/in/mehmet-ozan-kabak/)

Reviewer 2: ```suggestion
[Mehmet Ozan Kabak]: https://www.linkedin.com/in/mehmet-ozan-kabak/
```

[Synnada]: https://www.synnada.ai/
[called on the community to improve DataFusion’s performance]: https://github.com/apache/datafusion/issues/11442#issuecomment-2226834443

<img
src="{{ site.baseurl }}/img/clickbench-datafusion-43/perf-over-time.png"
width="100%"
class="img-responsive"
alt="ClickBench performance results over time for DataFusion"
/>

**Figure 2**: ClickBench performance improved over 30% between DataFusion 34
(released Dec 2023) and DataFusion 43 (released Nov 2024).

Like all good optimization efforts, ours took sustained effort as DataFusion ran

> out [single 2x performance improvements] several years ago. Working together our

Reviewer 3: ```suggestion
out of [single 2x performance improvements] several years ago. Working together our
```

community of engineers from around the world[^3] and all experience levels[^4]
pulled it off (check out [this discussion] to get a sense). It may be a "[hobo
sandwich]" [^5], but it is a tasty one!

[single 2x performance improvements]: https://www.influxdata.com/blog/aggregating-millions-groups-fast-apache-arrow-datafusion
[this discussion]: https://github.com/apache/datafusion/issues/12821
[hobo sandwich]: https://db.cs.cmu.edu/seminar2024/

Of course, most of these techniques have been implemented and described before,
but until now they were only available in proprietary systems such as
[Vertica](https://www.vertica.com/), [DataBricks
Photon](https://www.databricks.com/product/photon), or
[Snowflake](https://www.snowflake.com/en/) or in tightly integrated open source
systems such as [DuckDB](https://duckdb.org/) or

> [ClickHouse](https://clickhouse.com/) which were not designed to extend.

Reviewer 3: ```suggestion
[ClickHouse](https://clickhouse.com/) which were not designed to be extended.
```

([here](https://github.com/apache/arrow-rs/issues/5374)) and three major epics
([here](https://github.com/apache/datafusion/issues/10918),
[here](https://github.com/apache/datafusion/issues/11790) and
[here](https://github.com/apache/datafusion/issues/11752)) in DataFusion.

[Using StringView / German Style Strings to Make Queries Faster: Part 1]: https://www.influxdata.com/blog/faster-queries-with-stringview-part-one-influxdb/
[Part 2]: https://www.influxdata.com/blog/faster-queries-with-stringview-part-two-influxdb/

Here is a partial list of people involved in the project (I am sorry to those whom I forgot)

* **Arrow**:  [Xiangpeng Hao](https://github.com/XiangpengHao) (InfluxData’s amazing 2024 summer intern and UW Madison PhD), [Yijun Zhao](https://github.com/ariesdevil) from DataBend Labs, and [Raphael Taylor-Davies](https://github.com/tustvold) laid the foundation.  [RinChanNOW](https://github.com/RinChanNOWWW) from Tencent and [Andrew Duffy](https://github.com/a10y) from SpiralDB helped push it along in the early days, and [Liang-Chi Hsieh](https://github.com/viirya), [Daniël Heres](https://github.com/Dandandan) reviewed and provided guidance.

> * **DataFusion**:  [Xiangpeng Hao](https://github.com/XiangpengHao), again charted the initial path and [Alex Huang](https://github.com/Weijun-H), [Dharan Aditya](https://github.com/dharanad) [Lordworms](https://github.com/Lordworms), [Jax Liu](https://github.com/goldmedal),  [wiedld](https://github.com/wiedld), [Tai Le Manh](https://github.com/tlm365), [yi wang](https://github.com/my-vegetable-has-exploded), [doupache](https://github.com/doupache), [Jay Zhan](https://github.com/jayzhan211) , [Xin Li](https://github.com/xinlifoobar)  and [Kaifeng Zheng](https://github.com/Kev1n8) made it real.

Reviewer 4: ```suggestion
* **DataFusion**:  [Xiangpeng Hao](https://github.com/XiangpengHao), again charted the initial path and [Weijun Huang](https://github.com/Weijun-H), [Dharan Aditya](https://github.com/dharanad) [Lordworms](https://github.com/Lordworms), [Jax Liu](https://github.com/goldmedal),  [wiedld](https://github.com/wiedld), [Tai Le Manh](https://github.com/tlm365), [yi wang](https://github.com/my-vegetable-has-exploded), [doupache](https://github.com/doupache), [Jay Zhan](https://github.com/jayzhan211) , [Xin Li](https://github.com/xinlifoobar)  and [Kaifeng Zheng](https://github.com/Kev1n8) made it real.
```

* **DataFusion String Function Migration**:  [Trent Hauck](https://github.com/tshauck) organized the effort and set the patterns, [Jax Liu](https://github.com/goldmedal) made a clever testing framework, and [Austin Liu](https://github.com/austin362667), [Dmitrii Bu](https://github.com/demetribu), [Tai Le Manh](https://github.com/tlm365), [Chojan Shang](https://github.com/PsiACE), [WeblWabl](https://github.com/devanbenz), [Lordworms](https://github.com/Lordworms), [iamthinh](https://github.com/thinh2), [Bruce Ritchie](https://github.com/Omega359), [Kaifeng Zheng](https://github.com/Kev1n8), and [Xin Li](https://github.com/xinlifoobar) bashed out the conversions.

## Parquet

> Part DataFusion's speed in ClickBench is reading Parquet files (really) quickly,

Reviewer 5: ```suggestion
Part of the reason for DataFusion's speed in ClickBench is reading Parquet files (really) quickly,
```

which reflects invested effort in the Parquet reading system (see [Querying

> Parquet with Millisecond Latency]

Reviewer 3: ```suggestion
Parquet with Millisecond Latency] )
```

The [DataFusion ParquetExec] (built on the [Rust Parquet Implementation]) is now the most
sophisticated open source Parquet reader I know of. It has every optimization we
can think of for reading Parquet, including projection pushdown, predicate
pushdown (row group metadata, page index, and bloom filters), limit pushdown,
parallel reading, interleaved I/O, and late materialized filtering (coming soon ™️
by default). Some recent work from [June](https://github.com/itsjunetime)
[recently unblocked a remaining hurdle] for enabling late materialized
filtering, and conveniently [Xiangpeng Hao](https://github.com/XiangpengHao) is
working on the [final piece] (no pressure😅)

[recently unblocked a remaining hurdle]: https://github.com/apache/datafusion/pull/12135
[final piece]: https://github.com/apache/arrow-datafusion/issues/3463

[Querying Parquet with Millisecond Latency]: https://www.influxdata.com/blog/querying-parquet-millisecond-latency/
[DataFusion ParquetExec]: https://docs.rs/datafusion/latest/datafusion/datasource/physical_plan/parquet/struct.ParquetExec.html
[Rust Parquet Implementation]: https://crates.io/crates/parquet

> ## Skipping Partial Aggregation When Doesn't Help

Reviewer 5: **Skipping Partial Aggregation When Doesn't Help**

Is there a missing word in this title?

Author: ```suggestion
## Skipping Partial Aggregation When It Doesn't Help
```

Many ClickBench queries are aggregations that summarize millions of rows, a
common task for reporting and dashboarding. DataFusion uses state of the art
[two phase aggregation] plans. Normally, two phase aggregation works well as the
first phase consolidates many rows immediately after reading, while the data is
still in cache. However, for certain “high cardinality” aggregate queries (that
have large numbers of groups), [the two phase aggregation strategy used in
DataFusion was inefficient](https://github.com/apache/datafusion/issues/6937),
manifesting in relatively slower performance compared to other engines for
ClickBench queries such as

```sql
SELECT "WatchID", "ClientIP", COUNT(*) AS c, ... 
FROM hits 
GROUP BY "WatchID", "ClientIP" /* <----- 13M Distinct Groups!!! */
ORDER BY c DESC 
LIMIT 10;
```

> For such queries, the first first aggregation phase does not significantly

Reviewer 5: ```suggestion
For such queries, the first aggregation phase does not significantly
```

as shown in Figure 5. The cost of this copy [is especially high for variable
length strings and binary data].

[a general Row based mechanism]: https://github.com/apache/datafusion/blob/73507c307487708deb321e1ba4e0d302084ca27e/datafusion/physical-plan/src/aggregates/group_values/row.rs#L33-L39
[is especially high for variable length strings and binary data]: https://github.com/apache/datafusion/issues/9403

<img
src="{{ site.baseurl }}/img/clickbench-datafusion-43/row-based-storage.png"
width="100%"
class="img-responsive"
alt="Row based storage for multiple group columns"
/>

**Figure 5**: Prior to DataFusion 43.0.0, queries with multiple group columns
used Row based group storage and copied each group value twice. This copy
consumes a substantial amount of the query time for queries with many distinct
groups, such as several of the queries in ClickBench.

Many optimizations in Databases boil down to simply avoiding copies, and this
was no exception. The trick was to figure out how to avoid copies without
causing per-column comparison overhead to dominate or complexity to get out of
hand. In a great example of diligent and disciplined engineering, [Jay

> Zhan](http://jayzhan211) tried [several], [different] approaches until arriving

Reviewer 6: ```suggestion
Zhan](https://github.com/jayzhan211) tried [several], [different] approaches until arriving
```
